pkg/operator/sync.go

```diff
@@ -244,20 +244,50 @@ func (optr *Operator) syncRenderConfig(_ *renderConfig) error {
 	if err != nil {
 		return err
 	}
-	// as described by the name this is essentially static, but it no worse than what was here before.  Since changes disrupt workloads
-	// and since must perfectly match what the installer creates, this is effectively frozen in time.
-	initialKubeAPIServerServingCABytes, err := optr.getCAsFromConfigMap("openshift-config", "initial-kube-apiserver-server-ca", "ca-bundle.crt")
-	if err != nil {
+
+	bootstrapComplete := false
+	_, err = optr.clusterCmLister.ConfigMaps("kube-system").Get("bootstrap")
+	switch {
+	case err == nil:
+		bootstrapComplete = true
+	case apierrors.IsNotFound(err):
+		bootstrapComplete = false
+	case err != nil:
 		return err
+	default:
+		panic("math broke")
 	}
 
-	// Fetch the following configmap and merge into the the initial CA. The CA is the same for the first year, and will rotate
-	// automatically afterwards.
-	kubeAPIServerServingCABytes, err := optr.getCAsFromConfigMap("openshift-kube-apiserver-operator", "kube-apiserver-to-kubelet-client-ca", "ca-bundle.crt")
-	if err != nil {
-		kubeAPIServerServingCABytes = initialKubeAPIServerServingCABytes
+	var kubeAPIServerServingCABytes []byte
+	if bootstrapComplete {
+		// This is the ca-bundle published by the kube-apiserver that is used terminate client-certificates in the kube cluster.
+		// If the kubelet has a flag to check the in-cluster published ca bundle, that would be ideal.
+		kubeAPIServerServingCABytes, err = optr.getCAsFromConfigMap("openshift-config-managed", "kube-apiserver-client-ca", "ca-bundle.crt")
+		if err != nil {
+			return err
+		}
+
 	} else {
-		kubeAPIServerServingCABytes = mergeCertWithCABundle(initialKubeAPIServerServingCABytes, kubeAPIServerServingCABytes, "kube-apiserver-to-kubelet-signer")
+		// this else block exists to find a way to match the rendering hash of the machineconfig that the installer creates.
+		// it seems that if the MCO cannot find an existing machineconfig, the operator and machines perma-wedge and it cannot
```

Comment:
Could you help me understand a bit the exact race condition here? I feel like even with this else block, there are still potential conditions where we could trigger the not found wedge. And of course, I could be massively misunderstanding this, but reading this seems to imply that:

1. there is the proper "in-cluster CA" we fetch via `"openshift-config-managed", "kube-apiserver-client-ca", "ca-bundle.crt"`, which, if it doesn't exist,
2. there is the secondary location you can get (presumably the same?) CA at: `"openshift-kube-apiserver-operator", "kube-apiserver-to-kubelet-client-ca", "ca-bundle.crt"`, which, if it doesn't exist,
3. leaves us purely with the initial bootstrap CA `"openshift-config", "initial-kube-apiserver-server-ca", "ca-bundle.crt"`

Assuming 1 and 2 are the same thing here (maybe it isn't?) the only way the MCO should be able to mismatch is that the install time had `3` (just bootstrap) but then while the masters were coming up, something happened in parallel to generate `1`, such that the master nodes booted with `3` via the bootstrap MCS and the immediately was told that `1` was the correct configuration, which now confuses the master node MCDs.

In that case, I am not 100% sure why this would fix the problem, since we don't have something with "just `3`" which is what the MCD would want to prevent the wedge. Fundamentally, the MCO just wants to nothing to have changed in between the bootstrap MCC generating the initially served master config and the now-formed-control-plane's MCC generating those same config.

Reply:
The only correct value is "openshift-config-managed", "kube-apiserver-client-ca", "ca-bundle.crt".  That configmap doesn't exist and cannot exist during the installer run because the signing certificates are generated by operators.

The other two values are correct enough to get a cluster started, but are wrong.  However, if we use just the correct bundle, the installer and MCO don't agree on the hash for the machine config and the install never finishes.

The MCO appears to consistently start at least once before bootstrap completes (this makes sense since it has to start in order to provide the machineconfigs to the other nodes), and this creates the machineconfig the installer wants and later creates one that is correct.

I'm open to any fix, but this means exactly what we want it to mean.  If the MCO needs some special code to make it work more reliably, please assist and open a PR on top of this one.

Reply:
To follow up the discussion on slack, I think this will work to generate a "more correct" state generally. It sounds like we don't actually mismatch between the bootstrap and in-cluster config originally, instead the MCO was not reading the right ca. i.e. we're not trying to fix a race condition, we're trying to fix a "generate the correct bundle" condition, am I right in that?